sale_order_discount_formula/models/sale_order.py

```diff
@@ -0,0 +1,104 @@
+        #    a) \s* = any number of starting spaces
+        #    b) [-+] = a mandatory '+' or '-' symbol
+        #    c) \s* = any number of spaces
+        #    d) \d+ = a digit sequence of length at least 1
+        #    e) ([,.]\d+)? = an optional decimal part, composed of a '.' or ','
+        #       symbol followed by a digital sequence of length at least 1
+        #    f) \s* = any number of ending spaces
+
+        if discount and not discount_regex.match(discount):
+            return False
+        return True
+
+    @api.onchange('multiple_discount')
+    def onchange_multiple_discount(self):
+        def _normalize_discount(discount):
+            discount = discount.replace(" ", "")
+            discount = discount.replace(",", ".")
+            if discount and discount[0] == '+':
+                discount = discount[1:]
+            return discount
+
+        for line in self:
+            if line.multiple_discount:
+                if self._validate_discount(line.multiple_discount):
+                    normalized_discount = _normalize_discount(
+                        line.multiple_discount)
+                else:
+                    line.discount = 0
+                    raise Warning(
+                        _('Warning! The discount format is not recognized.'))
+
+                tokens = re.split(r'([+-])', normalized_discount)
+                numeric_tokens = []
+                last_sign = 1
+                for token in tokens:
+                    if token == '-':
+                        last_sign = -1
+                    elif token == '+':
+                        last_sign = 1
+                    else:
+                        numeric_tokens.append(float(token)*last_sign)
+                marginal_discount = 1
+                for token in numeric_tokens:
+                    marginal_discount = marginal_discount * (1-(token/100))
+                total_discount = 1 - marginal_discount
+                line.discount = total_discount * 100
+
+                if normalized_discount != line.multiple_discount:
+                    line.multiple_discount = normalized_discount
+
+            else:
+                line.discount = 0
+
+    @api.constrains('multiple_discount')
+    def validate_discount(self):
+        for line in self:
+            if line.multiple_discount and not self._validate_discount(
+                    line.multiple_discount):
+                raise ValidationError(
+                    _('Warning! The discount format is not recognized.'))
+
+    @api.multi
+    def write(self, vals):
+        res = super(SaleOrderLine, self).write(vals)
+        if 'multiple_discount' in vals:
+            for line in self:
+                line.onchange_multiple_discount()
+        return res
+
+    @api.multi
+    def _prepare_invoice_line(self, qty):
+        self.ensure_one()
+        res = super(SaleOrderLine, self)._prepare_invoice_line(qty)
+        res['multiple_discount'] = self.multiple_discount
```

Comment:
why do you put `multiple_discount` in here?

Reply:
Because _multiple_discount_ is also an attribute of _account.invoice.line_ that needs to be propagated from _sale.order.line_ when this method is invoked.